### PR TITLE
[Feature] Use new cusparse API to support CUDA 11.

### PR DESCRIPTION
**Pull request by yzh119**
## Description
CuSparse have deprecated its APIs `cusparseScsrmm2`/`cusparseDcsrmm2` and turn to use `cusparseSpMM`in CUDA 11, thus CUDA11 users could not build our repo because of the API breaking changes(#1969 ).
This PR replaces old API calls with new ones when CUDA version >=11 is detected. I've verified its correctness/performance on some examples such as GCN/GAT.

This is the first step towards utilizing CUDA 11, they also have some new features (e.g. SDDMM and float16), we can utilize them in later PRs.

This PR may not be a part of DGL v0.5, but users who are using CUDA 11 could try installing DGL from source on this branch.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: src/array/cuda/csr_transpose.cc [hunk 1]
```diff
@@ -47,7 +47,7 @@ CSRMatrix CSRTranspose<kDLGPU, int32_t>(CSRMatrix csr) {
   int32_t* t_indices_ptr = static_cast<int32_t*>(t_indices->data);
   void* t_data_ptr = t_data->data;
 
-#if __CUDA_API_VERSION >= 10010
+#if CUDART_VERSION >= 10010
```
Comment (yzh119) [inline src/array/cuda/csr_transpose.cc:50]: Previously this branch was not activated.

File: cmake/modules/CUDA.cmake [hunk 2]
```diff
@@ -43,15 +43,15 @@ set(CUDA_gpu_detect_output "")
       #find vcvarsall.bat and run it building msvc environment
       get_filename_component(MY_COMPILER_DIR ${CMAKE_CXX_COMPILER} DIRECTORY)
       find_file(MY_VCVARSALL_BAT vcvarsall.bat "${MY_COMPILER_DIR}/.." "${MY_COMPILER_DIR}/../..")
-      execute_process(COMMAND ${MY_VCVARSALL_BAT} && ${CUDA_NVCC_EXECUTABLE} -arch sm_30 --run  ${__cufile}
+      execute_process(COMMAND ${MY_VCVARSALL_BAT} && ${CUDA_NVCC_EXECUTABLE} -arch sm_35 --run  ${__cufile}
```
Comment (yzh119) [inline cmake/modules/CUDA.cmake:46]: sm30 is no longer supported.

File: CMakeLists.txt [hunk 3]
```diff
@@ -77,9 +77,9 @@ if(MSVC)
   endif()
 else(MSVC)
   include(CheckCXXCompilerFlag)
-  check_cxx_compiler_flag("-std=c++11"    SUPPORT_CXX11)
+  check_cxx_compiler_flag("-std=c++14"    SUPPORT_CXX14)
```
Comment (yzh119) [inline CMakeLists.txt:80]: @BarclayII I'm using the latest version of cuda-gcc (9.3.0) so I must set standard to c++14 to make it work.
Could we make it a variable depending on the compiler version?

Comment (BarclayII) [reply CMakeLists.txt:80]: This crashes on CentOS 7 where GCC 4.8.5 doesn't support C++14.

Does CUDA-GCC 9.3.0 really require C++14?  What was the error message it gave you if you use C++11?  My Fedora is using CUDA-GCC 8.3.0 and it worked fine with C++11 albeit without CUDA 11.

Comment (yzh119) [reply CMakeLists.txt:80]: Sorry, I have confirmed that C++14 is not necessarily required but thrust throws a warning encouraging me to compile DGL with C++11.
```
/opt/cuda/include/thrust/detail/config/cpp_dialect.h:104:13: warning: Thrust requires C++14. Please pass -std=c++14 to your compiler. Define THRUST_IGNORE_DEPRECATED_CPP_DIALECT to suppress this message.
```

Comment (BarclayII) [reply CMakeLists.txt:80]: I see.  So thrust requires C++14.  You can switch the flag by checking CMake Variable `CUDA_VERSION_MAJOR`.

File: src/kernel/cuda/binary_reduce_sum.cu [hunk 4]
```diff
@@ -125,6 +165,7 @@ void CusparseCsrmm2(
       static_cast<int32_t*>(csr.indptr->data),
       static_cast<int32_t*>(csr.indices->data),
       B_data, n, &beta, trans_out, m));
```
Comment (BarclayII) [inline src/kernel/cuda/binary_reduce_sum.cu:167]: missing `CUSPARSE_CALL(cusparseDestroyMatDescr(descr));`

**Review by BarclayII — APPROVED** [review]
LGTM

**Pull request merged**